x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -794,316 +775,294 @@ private void findTokenFromRefreshToken(String refreshToken, ActionListener<Searc
+     * token and refresh token are returned to the listener.
      */
     private void innerRefresh(String tokenDocId, Map<String, Object> source, long seqNo, long primaryTerm, Authentication clientAuth,
-                              ActionListener<Tuple<UserToken, String>> listener, Iterator<TimeValue> backoff, Instant refreshRequested) {
-        logger.debug("Attempting to refresh token [{}]", tokenDocId);
+                              Iterator<TimeValue> backoff, Instant refreshRequested, ActionListener<Tuple<UserToken, String>> listener) {
+        logger.debug("Attempting to refresh token stored in token document [{}]", tokenDocId);
         final Consumer<Exception> onFailure = ex -> listener.onFailure(traceLog("refresh token", tokenDocId, ex));
-        final Optional<ElasticsearchSecurityException> invalidSource = checkTokenDocForRefresh(source, clientAuth);
-        if (invalidSource.isPresent()) {
-            onFailure.accept(invalidSource.get());
-        } else {
-            if (eligibleForMultiRefresh(source, refreshRequested)) {
-                final Map<String, Object> refreshTokenSrc = (Map<String, Object>) source.get("refresh_token");
-                final String supersedingTokenDocId = (String) refreshTokenSrc.get("superseded_by");
-                logger.debug("Token document [{}] was recently refreshed, attempting to reuse [{}] for returning an " +
-                    "access token and refresh token", tokenDocId, supersedingTokenDocId);
-                final ActionListener<GetResponse> getSupersedingListener = new ActionListener<GetResponse>() {
-                    private final Consumer<Exception> maybeRetryOnFailure = ex -> {
-                        if (backoff.hasNext()) {
-                            final TimeValue backofTimeValue = backoff.next();
-                            logger.debug("retrying after [" + backofTimeValue + "] back off");
-                            final Runnable retryWithContextRunnable = client.threadPool().getThreadContext()
-                                    .preserveContext(() -> getTokenDocAsync(supersedingTokenDocId, this));
-                            client.threadPool().schedule(retryWithContextRunnable, backofTimeValue, GENERIC);
-                        } else {
-                            logger.warn("back off retries exhausted");
-                            onFailure.accept(ex);
-                        }
-                    };
+        final Tuple<RefreshTokenStatus, Optional<ElasticsearchSecurityException>> checkRefreshResult;
+        try {
+            checkRefreshResult = checkTokenDocumentForRefresh(clock.instant(), clientAuth, source);
+            if (checkRefreshResult.v2().isPresent()) {
+                onFailure.accept(checkRefreshResult.v2().get());
+                return;
+            }
+        } catch (IOException | DateTimeException | IllegalStateException e) {
```

Comment:
Which method throws a DateTimeException ? I think it's preferred to add it to the method signature even though it's a `RuntimeException` since we catch and handle it.

Reply:
It is `Instant.ofEpochMilli`. I have pushed f2da70363cd1bf6b53b60e7 .